File: 82. Remove Duplicates from Sorted List II/82. Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }
        
        ListNode dummy(0, head);
        ListNode* prev = &dummy;
        ListNode* current = head;
        
        while (current && current->next) {
            if (current->val == current->next->val) {
                int duplicateValue = current->val;
                while (current && current->val == duplicateValue) {
                    ListNode* temp = current;
                    current = current->next;
                    delete temp;
                }
                prev->next = current;
            } else {
                prev = current;
                current = current->next;
            }
        }
        
        return dummy.next;
    }
};

File: 82. Remove Duplicates from Sorted List II/82. Remove Duplicates from Sorted List II.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? deleteDuplicates(ListNode? head) {
    if (head == null || head.next == null) {
      return head;
    }
    
    final dummy = ListNode(0, head);
    ListNode? prev = dummy;
    ListNode? current = head;
    
    while (current != null && current.next != null) {
      if (current.val == current.next!.val) {
        int duplicateValue = current.val;
        while (current != null && current.val == duplicateValue) {
          current = current.next;
        }
        prev?.next = current;
      } else {
        prev = current;
        current = current.next;
      }
    }
    
    return dummy.next;
  }
}